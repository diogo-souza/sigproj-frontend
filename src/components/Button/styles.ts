import styled, { css, DefaultTheme } from 'styled-components';
import { darken } from 'polished';
import { ButtonProps } from '.';

type ContainerProps = {
  hasIcon: boolean;
} & Pick<ButtonProps, 'size' | 'color' | 'fullWidth'>;

const ContainerModifiers = {
  small: (theme: DefaultTheme) => css`
    height: 3rem;
    font-size: ${theme.font.sizes.xsmall};
  `,
  medium: (theme: DefaultTheme) => css`
    height: 4rem;
    font-size: ${theme.font.sizes.small};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.medium};
  `,
  large: (theme: DefaultTheme) => css`
    height: 5rem;
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.large};
  `,
  fullWidth: (theme: DefaultTheme) => css`
    width: 100%;
    span + svg {
      margin-left: ${theme.spacings.xxsmall};
    }
  `,
  withIcon: (theme: DefaultTheme) => css`
    span + svg {
      margin-left: ${theme.spacings.xxsmall};
    }
  `,
  disabled: () => css`
    &:disabled {
      cursor: not-allowed;
      filter: saturate(30%);
    }
  `,
};

export const Container = styled.button<ContainerProps>`
  ${({ theme, size, color, fullWidth, hasIcon, disabled }) => css`
    display: inline-flex;
    align-items: center;
    justify-content: center;
    text-decoration: none;
    background-color: ${theme.colors[color!]};
    color: ${theme.colors.titleWhite};
    border: 0;
    border-radius: ${theme.border.radius};
    padding: ${theme.spacings.xxsmall};

    &:focus {
      box-shadow: 0 0 0 3px ${theme.colors.focusShadow};
      border-color: ${theme.colors.focus};
    }

    ${hasIcon && ContainerModifiers.withIcon(theme)}
    ${!!size && ContainerModifiers[size](theme)};
    ${!!fullWidth && ContainerModifiers.fullWidth(theme)};
    ${disabled
      ? ContainerModifiers.disabled()
      : () => css`
          transition: background-color 0.2s;
          :hover {
            background-color: ${darken(0.1, theme.colors[color!])};
          }
        `}
  `}
`;
