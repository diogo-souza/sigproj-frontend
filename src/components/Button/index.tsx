import React, { AnchorHTMLAttributes, ButtonHTMLAttributes } from 'react';
import * as S from './styles';

type ButtonTypes =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>;

export type ButtonProps = {
  size?: 'small' | 'medium' | 'large';
  color?: 'primary' | 'secondary' | 'tertiary';
  fullWidth?: boolean;
  icon?: JSX.Element;
  as?: React.ElementType;
} & ButtonTypes;

const Button: React.FC<ButtonProps> = ({
  children,
  icon,
  size = 'medium',
  color = 'primary',
  fullWidth = false,
  ...props
}: ButtonProps) => (
  <S.Container
    size={size}
    color={color}
    fullWidth={fullWidth}
    hasIcon={!!icon}
    {...props}
  >
    {!!children && <span>{children}</span>}
    {!!icon && icon}
  </S.Container>
);

export default Button;
