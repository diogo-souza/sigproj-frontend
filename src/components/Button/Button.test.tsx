import { screen } from '@testing-library/react';
import { renderWithTheme } from 'utils/tests/helpers';
import { FaUser } from 'react-icons/fa';
import Button from '.';

describe('<Button />', () => {
  it('should render the medium size by default', () => {
    const { container } = renderWithTheme(<Button>Test</Button>);
    expect(screen.getByRole('button', { name: /Test/i })).toHaveStyle({
      height: '4rem',
      padding: '0.8rem 3.2rem',
      'font-size': '1.4rem',
    });

    expect(container.firstChild).toMatchSnapshot();
  });
  it('should render the small size', () => {
    renderWithTheme(<Button size="small">Test</Button>);
    expect(screen.getByRole('button', { name: /Test/i })).toHaveStyle({
      height: '3rem',
      'font-size': '1.2rem',
    });
  });
  it('should render the large size', () => {
    renderWithTheme(<Button size="large">Test</Button>);
    expect(screen.getByRole('button', { name: /Test/i })).toHaveStyle({
      height: '5rem',
      padding: '0.8rem 4.0rem',
      'font-size': '1.6rem',
    });
  });
  it('should render a fullWidth version', () => {
    renderWithTheme(<Button fullWidth>Test</Button>);
    expect(screen.getByRole('button', { name: /Test/i })).toHaveStyle({
      width: '100%',
    });
  });
  it('should render an icon version', () => {
    renderWithTheme(<Button icon={<FaUser data-testid="icon" />}>Test</Button>);
    expect(screen.getByText(/Test/i)).toBeInTheDocument();
    expect(screen.getByTestId(/icon/i)).toBeInTheDocument();
  });
});
