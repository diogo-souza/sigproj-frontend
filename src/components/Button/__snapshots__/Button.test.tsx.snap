// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the medium size by default 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #B51222;
  color: #FFFFFF;
  border: 0;
  border-radius: 0.4rem;
  padding: 0.8rem;
  height: 4rem;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
  -webkit-transition: background-color 0.2s;
  transition: background-color 0.2s;
}

.c0:focus {
  box-shadow: 0 0 0 3px rgba(75,174,227,0.4);
  border-color: #0d6efd;
}

.c0:hover {
  background-color: #870d19;
}

<button
  class="c0"
  color="primary"
>
  <span>
    Test
  </span>
</button>
`;
